server/src/instant/model/rule.clj

```diff
@@ -68,14 +69,34 @@
       conn
       ["DELETE FROM rules WHERE app_id = ?::uuid" app-id]))))
 
-(defn with-binds [rule etype expr]
-  (->> (get-in rule [etype "bind"])
-       (partition-all 2)
-       reverse
-       (reduce
-        (fn [body [var-name var-body]]
-          (str "cel.bind(" var-name ", " var-body ", " body ")"))
-        expr)))
+(defn bind-usages [compiler bind-keys expr]
+  (clojure.set/intersection bind-keys
+                            (cel/ident-usages compiler expr)))
+
+
+(defn with-binds [rule etype action expr]
+  (let [binds (get-in rule [etype "bind"])]
+    (if (empty? binds)
+      expr
+      (let [compiler (cel/action->compiler action)
+            bind-map (apply hash-map binds)
```

Comment:
This is really cool! 

One other win: 

Because of this change `binds` doesn't have to be an array anymore! It _could_ be a hash-map, which imo would make it a bit tighter visually! (Originally I chose the array so that we compiled 'in order', and skirted this)

Maybe down the road we can update our rule lang to accept bind as an object